[Templates] Plugin deps & Custom.d.ts

joshuaellis opened the pull request:
<!--
Hello 👋 Thank you for submitting a pull request.

To help us merge your PR, make sure to follow the instructions below:

- Create or update the tests
- Create or update the documentation at https://github.com/strapi/documentation
- Refer to the issue you are closing in the PR description: Fix #issue
- Specify if the PR is ready to be merged or work in progress (by opening a draft PR)

Please ensure you read the Contributing Guide: https://github.com/strapi/strapi/blob/main/CONTRIBUTING.md
-->

### What does it do?

* Correctly adds the deps that are listed in the generated plugin template
* Correctly adds the dev deps types packages in the generated plugin template for TS only
* Adds `custom.d.ts` file for Typescript plugins so it doesn't fail with our own packages

### Why is it needed?

* When you try to make a plugin with those template, they fail horrifically because we don't list most of the deps. We assume they're in a monorepo set up but that isn't necessarily the case and is poor practice imo.

### Related issue(s)/PR(s)

* resolves #15741
* relates to #15072

markkaylor commented on packages/generators/generators/lib/templates/js/plugin-package.json.hbs:
```diff
@@ -8,7 +8,25 @@
   },
-  "dependencies": {},
+  "dependencies": {
+    "@strapi/design-system": "^1.6.3",
+    "@strapi/helper-plugin": "^4.6.0",
+    "@strapi/icons": "^1.6.3",
```
Do we have to update these manually after each release? Same concern / question for the other dependencies.

joshuaellis replied:
IMHO – no, the `^` will mean it will install the most recent version except for breaking changes – IF someone needs to lock a version they can do that themselves & we resolve the deps to our required version at the moment.

markkaylor replied:
In this `.hbs` file they will update too 🤔 ?

joshuaellis replied:
> when it is installed in the plugin it will fetch the most recent version

Yes, sorry i should have been clearer.

markkaylor commented on packages/generators/generators/lib/files/ts/plugin/tsconfig.json:
```diff
@@ -1,24 +1,19 @@
     "node_modules/",
     "dist/",
 
     // Do not include admin files in the server compilation
```
I think this comment may need to be updated or removed 

```suggestion
    // Do not include server files in the server compilation
```

markkaylor commented on packages/generators/generators/lib/templates/ts/plugin-package.json.hbs:
```diff
@@ -8,10 +8,29 @@
+    "react-router-dom": "^5.3.4",
+    "styled-components": "^5.3.6",
     "typescript": "4.6.3"
+  },
+    "peerDependencies": {
```
```suggestion
  },
  "peerDependencies": {
```

gu-stav commented on packages/generators/generators/lib/files/ts/plugin/tsconfig.server.json:
```diff
@@ -0,0 +1,25 @@
+    // Include the root directory
+    "server",
+    // Force the JSON files in the src folder to be included
+    "server/**/*.json"
```
As far as I remember these paths have created quite a few problems on windows machines. @Convly could you have a look?

joshuaellis replied:
These paths aren't different from what they are now? and afaik TS handles these internally regardless of how you write them?

gu-stav replied:
I don't know anything about generators and the way they configure TS - just remembered there were problems at some point. 🤷🏼

Convly replied:
Internally TS automatically parse `/` and `\` 
https://github.com/microsoft/TypeScript/blob/main/src/compiler/path.ts
So _in theory_ we should not have an issue

gu-stav commented on packages/generators/generators/lib/templates/js/plugin-package.json.hbs:
```diff
@@ -8,7 +8,25 @@
+    "prop-types": "^15.7.2"
+  },
+  "devDependencies": {
+    "react": "^17.0.2",
+    "react-dom": "^17.0.2",
+    "react-router-dom": "^5.3.4",
+    "styled-components": "^5.3.6"
```
Why are these dev-dependencies?

joshuaellis replied:
Because if someone publishes a plugin they shouldn't be installing their own versions of these libs they should be peer deps otherwise it creates errors like "Cannot find X", especially with react-router-dom.

While I _know_ the current plugin setup wouldn't let this happen (because we force resolution), we should be pre-emptive and support best practice here imo

gu-stav replied:
Right, that makes sense. Thanks!

gu-stav commented:
@joshuaellis Unfortunately I don't feel comfortable to review this PR. Generators are owned by @markkaylor and @remidej and I believe they should be the ones reviewing this.

markkaylor reviewed (commented):
When I do the following test for a TypeScript plugin:

1. Generate a TypeScript plugin via `./node_modules/.bin/strapi generate` in examples/getstarted
2. Add the plugin config
3. In the TypeScript plugin run yarn build

The same test seems to work on `main` though 🤔 

<img width="1428" alt="Screenshot 2023-02-16 at 16 57 07" src="https://user-images.githubusercontent.com/26598053/219352666-a75bfce0-e8dc-401b-9558-ca039f0479b7.png">

joshuaellis commented:
Should we remove `axiosInstance` considering we now have `useFetchClient` hook? @gu-stav  & @markkaylor

gu-stav commented:
@joshuaellis I'd love that.

joshuaellis commented:
So because we don't have types for the `StrapiApp` class, i've had to put `any`. I've also removed the `react` deprecated classes because imo inference is easier with react-components -> `React.FC` in 17.X was highly discouraged to be used.

I've also removed `axiosInstance` because people should be using `useFetchClient` – I couldn't see any documentation on the site for this though... so i'm not sure how we communicate this change?

gu-stav commented on packages/generators/generators/lib/templates/ts/plugin-package.json.hbs:
```diff
@@ -44,8 +44,8 @@
     "npm": ">=6.0.0"
   },
   "scripts": {
-    "develop": "tsc -w",
```
I'd rather keep this as `develop`, because this is what we are using everywhere.

markkaylor commented on examples/getstarted/src/index.js:
```diff
@@ -7,7 +7,9 @@ module.exports = {
    */
-  register({ strapi }) {},
+  register(app) {
+    console.log(app);
```
Just need to get rid of this guy 
```suggestion

```

markkaylor reviewed (approved):
Looks good to me 👍 

> I couldn't see any documentation on the site for this though... so i'm not sure how we communicate this change?

I guess maybe we could add a small comment in the release notes?

markkaylor commented on examples/getstarted/src/index.js:
```diff
@@ -7,7 +7,8 @@ module.exports = {
    * This gives you an opportunity to extend code.
    */
-  register({ strapi }) {},
+  register(app) {
+  },
```
Sorry but this will keep me up at night
```suggestion
  register(app) {},
```

markkaylor commented on examples/getstarted/src/index.js:
```diff
@@ -7,7 +7,7 @@ module.exports = {
    * This gives you an opportunity to extend code.
    */
-  register({ strapi }) {},
+  register(app) {},
```
Why are we no longer destructuring strapi here? I'm not sure we should have a diff at all in this file

joshuaellis replied:
thanks for highlighting that, i was using the github commit system so wasn't checking well enough!
